### ISSUE TITLE
`iris.tests.stock.netcdf` excessive memory usage with unlimited dimensions

### ISSUE THREAD

**Issue by trexfeathers** [SciTools/iris#4834]
## TL;DR

Using [`iris.tests.stock.netcdf._add_standard_data()`](https://github.com/SciTools/iris/blob/a908a27937e4ca8eb810001ebd64f11ff75dce72/lib/iris/tests/stock/netcdf.py#L95-L135) demands much more memory than expected for variables spanning both **unlimited** and standard dimensions, only with NetCDF-4 format files.

## Context

[`iris.tests.stock`](https://github.com/SciTools/iris/tree/a908a27937e4ca8eb810001ebd64f11ff75dce72/lib/iris/tests/stock) includes the ability to generate NetCDF files based on CDL header templates. Here are the steps:

1. Populate CDL template placeholders using Python's `string.Template`
2. Create a 'dataless' NetCDF file by using `ncgen` on the CDL header string
3. Add data to the NetCDF file using the `netcdf4-python` module

## Detail

@pp-mo and @trexfeathers have discovered that adding data for variables that span both **unlimited** and standard dimensions demands memory >100 times greater than the array's `nbytes` property. This is not true if the variable exclusively spans an unlimited dimension, or spans no unlimited dimensions. We have determined this by creating various CDL templates then stepping through [`iris.tests.stock.netcdf._add_standard_data()`](https://github.com/SciTools/iris/blob/a908a27937e4ca8eb810001ebd64f11ff75dce72/lib/iris/tests/stock/netcdf.py#L95-L135).

This problem is only seen when we use `ncgen` to create NetCDF-4 format files. NetCDF-3 format files are OK. This suggests a problem related to HDF5.

We use the steps above so that it is easy to customise the size of the files dimensions, and to add appropriately formatted data (e.g. strictly monotonic arrays for dimension coordinates). However it's probably a strange pattern unforeseen by the developers of `ncgen` nor `netcdf4-python`, so we could have tripped over something unexpected. This also makes the problem difficult for us to research using search engines, made no easier by the niche-ness of what we're doing.

## Next steps

* Beware the problem, sizing artificial files to avoid blowing memory during generation.
* See if the problem can be replicated without `netcdf4-python` - so sticking to `ncgen` and perhaps `nco`.
  This would help us work out which layer the problem resides in.
